platform/darwin/src/MGLNetworkConfiguration.h

```diff
@@ -4,13 +4,39 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+@class MGLNetworkConfiguration;
+
+/**
+ :nodoc:
+ The `MGLNetworkConfigurationSessionDelegate` protocol provides an `NSURLSession`
+ object for the `MGLNetworkConfiguration`. This API should be considered
+ experimental, likely to be removed or changed in future releases.
+
+ This method is called from background threads, i.e. it is not called on the main
+ thread.
+
+ @note: Background sessions (i.e. created with
+ `-[NSURLSessionConfiguration backgroundSessionConfigurationWithIdentifier:]`)
+ are not supported at this time.
+ */
+@protocol MGLNetworkConfigurationSessionDelegate <NSObject>
+- (NSURLSession *)sessionForNetworkConfiguration:(MGLNetworkConfiguration *)configuration;
+@end
+
+
 /**
  The `MGLNetworkConfiguration` object provides a global way to set a base
  `NSURLSessionConfiguration`, and other resources.
  */
 MGL_EXPORT
 @interface MGLNetworkConfiguration : NSObject
 
+/**
+ :nodoc:
+ Delegate for the `MGLNetworkConfiguration` class.
+ */
+@property (nonatomic, weak) id delegate;
```

Comment:
```suggestion
@property (nonatomic, weak) id <MGLNetworkConfigurationSessionDelegate> delegate;
```

Reply:
This is deliberately left as `id` for a couple of reasons:
- to keep this clearly undocumented
- so that delegates can optionally conform to `MGLNetworkConfigurationSessionDelegate` (rather than using `@optional` methods and hitting Swift issues). Perhaps there's a better way here?

Reply:
> to keep this clearly undocumented

Neither `delegate` nor `MGLNetworkConfigurationSessionDelegate` will become documented by virtue of using a strong type. On the other hand, we lose helpful compiler warnings when we misuse `delegate` internally.

> so that delegates can optionally conform to `MGLNetworkConfigurationSessionDelegate` (rather than using `@optional` methods and hitting Swift issues). Perhaps there's a better way here?

What Swift issues are you referring to? Swift supports optional methods in Objective-C protocols. We should make this protocol’s methods optional and require the delegate to be of the right type.

Reply:
> What Swift issues are you referring to?

Ah, I was thinking of the problems with optional properties & Swift. I've updated the type, and marked the method as optional.